### PR TITLE
SNOW-344502 - Improved performance when downloading large amounts of data.

### PR DESCRIPTION
**Pull request by stegus64**
* Uses handwritten parsers for integers and decimals
* Keeps data in UTF8 format until we absolutely need to convert it to strings

In a testcase that downloads 266 million rows we had the following runtime before this change:

.Net Framework 00:13:16
Core           00:05:16

After this change the same testcase has the following runtime:

.Net Framework 00:02:18
Core           00:02:29

### CODE REVIEW

**Comment by stegus64** [comment]
@sfc-gh-abhatnagar Have you seen this PR? 

I am not sure about the proper process to suggest improvements, but this PR will improve performance in certain important situations by a factor of 5 or more.

Are you not interested?

**Comment by SimbaGithub** [comment]
This will definitely improve conversion of numeric and date-time data!

File: Snowflake.Data/Core/FastParser.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Snowflake.Data.Core
+{
+    public class FastParser
+    {
+
+        public static Int64 FastParseInt64(byte[] s, int offset, int len)
```
Comment (SimbaGithub) [inline Snowflake.Data/Core/FastParser.cs:10]: What about adding _private static Int64 FastParseInt(byte[] s, int offset, int len, UInt64 maxValue)_, with the same code/logic than _FastParseInt64_ and _FastParseInt32_ and have them call FastParseInt with the correct max value? FastParseInt16 and FastParseByte could also be added easily and catch overflow ahead of FastParseInt32.

File: Snowflake.Data.Tests/SFDataConverterTest.cs [hunk 2]
```diff
@@ -97,5 +96,29 @@ private void internalTestConvertDate(DateTime dtExpected, DateTime testValue)
...
+        [Test]
+        [TestCase("9223372036854775807.9223372036854775807")]
+        [TestCase("-9223372036854775807.1234567890")]
+        [TestCase("-1.300")]
+        [TestCase("999999999999999999.000000000000100000000000")]
+        [TestCase("4294967295.4294967296")]
+        public void TestConvertToDecimal(string s)
+        {
+            decimal actual = (decimal)SFDataConverter.ConvertToCSharpVal(s, SFDataType.FIXED, typeof(decimal));
+            decimal expected = Convert.ToDecimal(s, System.Globalization.CultureInfo.InvariantCulture);
```
Comment (SimbaGithub) [inline Snowflake.Data.Tests/SFDataConverterTest.cs:120]:
```suggestion
            decimal expected = Convert.ToDecimal(s, CultureInfo.InvariantCulture);

```

File: Snowflake.Data/Core/FastParser.cs (hunk 1 above)
Comment (stegus64) [reply Snowflake.Data/Core/FastParser.cs:10]: There is a good reason for having separate implementations of FastParseInt32 and FastParseInt64. 
FastParseInt32 is faster because it only works with 32-bit integers. When the end user calls DataReader.GetInt(i) we know he only expects a 32-bit value so we can call the faster version FastParseInt32 directly.
I don't see a good reason to define a FastParseInt method.

**Comment by stegus64** [comment]
This PR also solves #273  #246  #219

File: Snowflake.Data/Core/SFDataConverter.cs [hunk 3]
```diff
@@ -20,74 +20,100 @@ public enum SFDataType
...
+                    return srcVal.Buffer[0] == '1' || srcVal.Buffer[0] == 't' || srcVal.Buffer[0] == 'T';
+                }
+                else if (destType == typeof(byte[]))
+                {
+                    return srcType == SFDataType.BINARY ?
+                        HexToBytes(srcVal.ToString()) : srcVal.GetBytes();
+                }
+                else if (destType == typeof(Int16))
+                {
+                    return (Int16)FastParser.FastParseInt32(srcVal.Buffer, srcVal.offset, srcVal.length);
```
Comment (SimbaGithub) [inline Snowflake.Data/Core/SFDataConverter.cs:82]: This will not catch overflows and return erroneous data to the user instead of an error.

File: Snowflake.Data/Core/SFDataConverter.cs [hunk 4]
```diff
@@ -20,74 +20,100 @@ public enum SFDataType
...
+                }
+                else if (destType == typeof(byte))
+                {
+                    return (byte)FastParser.FastParseInt32(srcVal.Buffer, srcVal.offset, srcVal.length);
```
Comment (SimbaGithub) [inline Snowflake.Data/Core/SFDataConverter.cs:86]: Here as well, it will also not catch overflows

File: Snowflake.Data/Core/FastParser.cs (hunk 1 above)
Comment (SimbaGithub) [reply Snowflake.Data/Core/FastParser.cs:10]: I am not sure how much this would affect the performance, but if this is the reason behind those almost 2 identical functions then it's fine.

Comment (stegus64) [reply Snowflake.Data/Core/FastParser.cs:10]: image

This is a benchmark of different parsing methods. As you can see FastParseInt32 is about 30-40% faster than FastParseInt64. I think it is well worth the effort to enable that speedup.

TypeConverter was the old code before this PR. As you can see, the old code was 57 times slower than the new version on .Net Framework

File: Snowflake.Data/Core/SFDataConverter.cs (hunk 3 above)
Comment (stegus64) [reply Snowflake.Data/Core/SFDataConverter.cs]: Good catch! I thought a narrowing cast generated overflow, but you are correct.
Fixed

File: Snowflake.Data/Core/SFDataConverter.cs (hunk 4 above)
Comment (stegus64) [reply Snowflake.Data/Core/SFDataConverter.cs]: Fixed

**Comment by SimbaGithub** [comment]
LGTM :)

@sfc-gh-hyu  : Can you take a look and see if you are Ok to green light this as well?

**Review by sfc-gh-hyu — COMMENTED** [review]
@SimbaGithub I understands the big picture but did not look into the detail. I think the change make sense to me. My only concern is that do we want to add some kind of parameter/config to fall back in case of we found some bugs in the new code?

**Comment by SimbaGithub** [comment]
@sfc-gh-hyu : We can always revert if needed, I think it's probably better than leaving both alive in the connector. Now I agree that this touches data and is crucial to the connector. I would expect that if all the tests pass then it should be Ok, but I don't have an extensive knowledge of the existing tests. Do you know of missing edge case tests which need to be added? Do you have something specific in mind?

I'll take another look for edge cases as well.

**Comment by SimbaGithub** [comment]
Here is the plan : When CI is back to normal, let's review the test cases, add more if missing ones are identified and then we should be good to go with integrating this in next month release.

**Comment by stegus64** [comment]
Sounds good

**Review by sfc-gh-hyu — APPROVED** [review]

File: Snowflake.Data.Tests/SFDataConverterTest.cs [hunk 5]
```diff
@@ -97,5 +96,96 @@ private void internalTestConvertDate(DateTime dtExpected, DateTime testValue)
...
+        [TestCase("-1.300")]
+        [TestCase("999999999999999999.000000000000100000000000")]
+        [TestCase("4294967295.4294967296")]
+        public void TestConvertToDecimal(string s)
+        {
+            decimal actual = (decimal)SFDataConverter.ConvertToCSharpVal(s, SFDataType.FIXED, typeof(decimal));
+            decimal expected = Convert.ToDecimal(s, CultureInfo.InvariantCulture);
+
+            Assert.AreEqual(expected, actual);
+        }
```
Comment (SimbaGithub) [inline Snowflake.Data.Tests/SFDataConverterTest.cs:189]: Can we had a few extra tests unsupported conversions and a couple of special cases?
```     }

        [Test]
        [TestCase("9223372036854775807.9223372036854775807")]
        [TestCase("-9223372036854775807.1234567890")]
        [TestCase("-1.300")]
        [TestCase("999999999999999999.000000000000100000000000")]
        [TestCase("4294967295.4294967296")]
        [TestCase("1.5e-36")]
        [TestCase("1.5e+38")]
        //[TestCase("inf")] -- TODO - Not supported yet
        //[TestCase("-inf")] -- TODO - Not supported yet
        [TestCase("NaN")]
        public void TestConvertToFloat(string s)
        {
            double actualDouble = (double)SFDataConverter.ConvertToCSharpVal(s, SFDataType.FIXED, typeof(double));
            double expectedDoulbe = Convert.ToDouble(s, CultureInfo.InvariantCulture);

            Assert.AreEqual(actualDouble, expectedDoulbe);

            float actualFloat = (float)SFDataConverter.ConvertToCSharpVal(s, SFDataType.FIXED, typeof(float));
            float expectedFloat = Convert.ToSingle(s, CultureInfo.InvariantCulture);

            Assert.AreEqual(expectedFloat, actualFloat);
        }

        [Test]
        [TestCase("thisIsNotAValidValue")]
        [TestCase("-1.300")]
        [TestCase("425.426")]
        [TestCase("1.5e-36")]
        [TestCase("1.5e+38")]
        [TestCase("inf")]
        [TestCase("-inf")]
        [TestCase("NaN")]
        public void TestInvalidConversionInvalidInt(string s)
        {
            Assert.Throws<FormatException>(() => SFDataConverter.ConvertToCSharpVal(s, SFDataType.FIXED, typeof(Int32)));
            Assert.Throws<FormatException>(() => SFDataConverter.ConvertToCSharpVal(s, SFDataType.FIXED, typeof(Int64)));
            Assert.Throws<FormatException>(() => SFDataConverter.ConvertToCSharpVal(s, SFDataType.FIXED, typeof(Int16)));
            Assert.Throws<FormatException>(() => SFDataConverter.ConvertToCSharpVal(s, SFDataType.FIXED, typeof(byte)));
        }

        [Test]
        [TestCase("thisIsNotAValidValue")]
        public void TestInvalidConversionInvalidFloat(string s)
        {
            Assert.Throws<FormatException>(() => SFDataConverter.ConvertToCSharpVal(s, SFDataType.FIXED, typeof(float)));
            Assert.Throws<FormatException>(() => SFDataConverter.ConvertToCSharpVal(s, SFDataType.FIXED, typeof(double)));
        }

        [Test]
        [TestCase("thisIsNotAValidValue")]
        [TestCase("1.5e-36")]
        [TestCase("1.5e+38")]
        [TestCase("inf")]
        [TestCase("-inf")]
        [TestCase("NaN")]
        public void TestInvalidConversionInvalidDecimal(string s)
        {
            Assert.Throws<FormatException>(() => SFDataConverter.ConvertToCSharpVal(s, SFDataType.FIXED, typeof(decimal)));
        }

```

Comment (stegus64) [reply Snowflake.Data.Tests/SFDataConverterTest.cs:189]: Added these tests to the PR

**Comment by SimbaGithub** [comment]
Thank you Stefan for the work on this! Going to merge it now. It will be in the next release.
